internal/pulsar/pulsar.go

```diff
@@ -0,0 +1,137 @@
+package pulsar
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	pulsarLog "github.com/apache/pulsar-client-go/pulsar/log"
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+type ClientConf struct {
+	url               string
+	operationTimeout  time.Duration
+	connectionTimeout time.Duration
+}
+
+type ProducerConf struct {
+	topic string
+}
+
+type Producer struct {
+	pulsar.Producer
+}
+
+type ProducerAdapter interface {
+	SendMessage(ctx context.Context, key, orderingKey string, msg []byte) error
+	SendMessageAsync(ctx context.Context, key, orderingKey string, msg []byte, statusFunc func(id pulsar.MessageID, message *pulsar.ProducerMessage, err error))
+	Close()
+	Flush() error
+}
+
+type Client struct {
+	pulsar.Client
+}
+
+func NewClient(config *config.Config) (Client, error) {
+	log := logger.NewLogger().Child("pulsar")
+	client, err := newPulsarClient(getClientConf(config), log)
+	if err != nil {
+		return Client{}, fmt.Errorf("error creating pulsar client : %v", err)
+	}
+	return client, nil
+}
+
+func (c *Client) NewProducer(config *config.Config) (ProducerAdapter, error) {
```

Comment:
It is possible to have a single pulsar client (using the same pulsar cluster for different use cases). But for every use case we are going to need to use different topics and/or producer/consumer configurations.

E.g. for event schemas we want our producer to batch using `BatcherBuilderType: pulsar.KeyBasedBatchBuilder`. This might not stand true for another producer operating against a different topic with different requirements.

Thus, I believe that NewProducer should accept `pulsar.ProducerOptions` instead of `config.Config`.